File: src/main/java/ting/leetcode/java/binarytree/BinaryTreePreorderTraversal.java
package ting.leetcode.java.binarytree;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class BinaryTreePreorderTraversal {
    public List<Integer> preorderTraversal(TreeNode root) {
        if (root == null) {
            return Collections.emptyList();
        }

        ArrayList<Integer> record = new ArrayList<>();
        record.add(root.val);
        record.addAll(preorderTraversal(root.left));
        record.addAll(preorderTraversal(root.right));
        return record;
    }
}

File: src/main/kotlin/ting/leetcode/tree/BinaryTreePreorderTraversal.kt
package ting.leetcode.tree

class BinaryTreePreorderTraversal {
    fun preorderTraversal(root: TreeNode?): List<Int> {
        if (root == null) return emptyList()

        val result = mutableListOf<Int>()
        val stack = ArrayDeque<TreeNode?>()

        stack.addFirst(root)
        while (stack.isNotEmpty()) {
            stack.run {
                val node = removeFirst()
                result.add(node?.`val`!!)
                if (node.right != null) addFirst(node.right)
                if (node.left != null) addFirst(node.left)
            }
        }

        return result
    }

    fun preorderTraversalRecursion(root: TreeNode?): List<Int> =
        if (root == null)
            emptyList()
        else
            buildList {
                add(root.`val`)
                addAll(preorderTraversalRecursion(root.left))
                addAll(preorderTraversalRecursion(root.right))
            }

}
